File: frontend/src/api/shop.ts
import { getAuthHeaders } from "./authHeaders";
import { apiRequest } from "./http";

export type QuotaType = "unlimited" | "personal_cap" | "auto_refill";

export type ItemType = "coin_pack" | "streak_freeze" | "time_treasure" | "revival";

export type TierRequired = "free" | "super";

export type ShopOffer = {
  id: string;
  item_code: string;
  display_name: string;
  rarity: string;
  price_asset_code: string;
  price_amount: number;
  is_active: boolean;
  active_from: string | null;
  active_to: string | null;
  purchase_limit_per_user: number | null;
  quota_type: QuotaType;
  max_capacity: number | null;
  refill_days: number | null;
  tier_required: TierRequired;
  experiment_flag: string | null;
  item_type: ItemType;
};

export type ShopBalance = {
  asset_code: string;
  balance: number;
};

export type ShopLedgerEntry = {
  id: string;
  asset_code: string;
  delta: number;
  balance_after: number;
  reason_code: string;
  source_type: string | null;
  source_id: string | null;
  created_at: string;
};

export type PurchaseShopItemInput = {
  offerId: string;
  idempotencyKey?: string;
};

export type PurchaseShopItemResponse = {
  id: string;
  user_id: string;
  offer_id: string | null;
  item_code: string;
  price_asset_code: string;
  price_amount: number;
  status: "pending" | "completed" | "failed" | "refunded";
  purchased_at: string;
};

export type ShopInventoryItem = {
  user_id: string;
  item_code: string;
  quantity: number;
  updated_at: string;
  quota_max: number | null;
  refill_days: number | null;
  last_refill_at: string | null;
  next_refill_at: string | null;
};

export type ShopInventoryResponse = {
  items: ShopInventoryItem[];
};

export type ListShopItemsParams = {
  userTier?: TierRequired;
  experimentFlags?: string[];
};

export const listShopItems = async (params: ListShopItemsParams = {}): Promise<ShopOffer[]> => {
  const query = new URLSearchParams();
  if (params.userTier) {
    query.set("user_tier", params.userTier);
  }
  if (params.experimentFlags && params.experimentFlags.length > 0) {
    query.set("experiment_flags", params.experimentFlags.join(","));
  }
  const qs = query.toString();
  const path = qs ? `/api/v1/shop/items?${qs}` : "/api/v1/shop/items";
  return apiRequest<ShopOffer[]>(path);
};

export const getShopBalance = async (): Promise<ShopBalance> => {
  return apiRequest<ShopBalance>("/api/v1/shop/balance", {
    headers: getAuthHeaders(),
  });
};

export const purchaseShopItem = async (
  input: PurchaseShopItemInput,
): Promise<PurchaseShopItemResponse> => {
  return apiRequest<PurchaseShopItemResponse>("/api/v1/shop/purchase", {
    method: "POST",
    headers: getAuthHeaders(),
    body: {
      offer_id: input.offerId,
      idempotency_key: input.idempotencyKey,
    },
  });
};

export const getShopInventory = async (): Promise<ShopInventoryItem[]> => {
  const response = await apiRequest<ShopInventoryResponse>("/api/v1/shop/inventory", {
    headers: getAuthHeaders(),
  });
  return response.items;
};

export const getShopLedger = async (): Promise<ShopLedgerEntry[]> => {
  return apiRequest<ShopLedgerEntry[]>("/api/v1/shop/ledger", {
    headers: getAuthHeaders(),
  });
};

File: backend/app/schemas/shop.py
from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.db.models.economy import PurchaseStatus


class QuotaType(StrEnum):
    UNLIMITED = "unlimited"
    PERSONAL_CAP = "personal_cap"
    AUTO_REFILL = "auto_refill"


class ItemType(StrEnum):
    COIN_PACK = "coin_pack"
    STREAK_FREEZE = "streak_freeze"
    XP_BOOST = "xp_boost"
    TIME_TREASURE = "time_treasure"
    REVIVAL = "revival"


class TierRequired(StrEnum):
    FREE = "free"
    SUPER = "super"


class ShopOfferResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    item_code: str
    display_name: str
    rarity: str
    price_asset_code: str
    price_amount: int
    is_active: bool
    active_from: datetime | None
    active_to: datetime | None
    purchase_limit_per_user: int | None
    # Quota / conditional access fields
    quota_type: QuotaType = QuotaType.UNLIMITED
    max_capacity: int | None = None
    refill_days: int | None = None
    tier_required: TierRequired = TierRequired.FREE
    experiment_flag: str | None = None
    item_type: ItemType = ItemType.COIN_PACK


class PurchaseRequest(BaseModel):
    offer_id: UUID
    idempotency_key: str | None = None


class PurchaseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    offer_id: UUID | None
    item_code: str
    price_asset_code: str
    price_amount: int
    status: PurchaseStatus
    purchased_at: datetime


class BalanceResponse(BaseModel):
    asset_code: str
    balance: int


class InventoryItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: UUID
    item_code: str
    quantity: int
    updated_at: datetime
    quota_max: int | None = None
    refill_days: int | None = None
    last_refill_at: datetime | None = None
    next_refill_at: datetime | None = None


class InventoryResponse(BaseModel):
    items: list[InventoryItemResponse]


class LeaderboardEntryResponse(BaseModel):
    user_id: UUID
    display_name: str | None
    total_xp: int
    rank: int


class UseItemRequest(BaseModel):
    item_code: str
    context: dict[str, Any] | None = None


class UseItemResponse(BaseModel):
    success: bool
    item_code: str
    quantity_remaining: int
    effect_applied: dict[str, Any] | None = None


class ActiveEffect(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    effect_type: str
    multiplier: float | None = None
    expires_at: datetime | None = None
    source_item_code: str | None = None
    context: dict[str, Any] | None = None
    created_at: datetime


class ActiveEffectsResponse(BaseModel):
    effects: list[ActiveEffect]
